### PR TITLE
[Flang][OpenMP] Port OpenMP single tests to hlfir flow

### PR DESCRIPTION
**Pull request by harishch4**

### CODE REVIEW

File: flang/test/Lower/OpenMP/single.f90 [hunk 1]
```diff
@@ -9,13 +9,14 @@
 !CHECK-SAME: (%[[x:.*]]: !fir.ref<i32> {fir.bindc_name = "x"})
 subroutine omp_single(x)
   integer, intent(inout) :: x
+  !CHECK: %[[X_DECL:.*]]:2 = hlfir.declare %[[x]] {fortran_attrs = #fir.var_attrs<intent_inout>, uniq_name = "_QFomp_singleEx"} : (!fir.ref<i32>) -> (!fir.ref<i32>, !fir.ref<i32>)
   !CHECK: omp.parallel
   !$omp parallel
   !CHECK: omp.single
   !$omp single
-    !CHECK: %[[xval:.*]] = fir.load %[[x]] : !fir.ref<i32>
+  !CHECK: %[[xval:.*]] = fir.load %[[X_DECL]]#0 : !fir.ref<i32>
```
Comment (kiranchandramohan) [inline flang/test/Lower/OpenMP/single.f90:17]: Nit: keep the alignment.

File: flang/test/Lower/OpenMP/single.f90 [hunk 2]
```diff
@@ -70,14 +72,18 @@ end subroutine single_allocate
...
+! CHECK:           %[[VAL_0_DECL:.*]]:2 = hlfir.declare %[[VAL_0]] {uniq_name = "_QFsingle_privatizationEx"} : (!fir.ref<f32>) -> (!fir.ref<f32>, !fir.ref<f32>)
+! CHECK:           %[[VAL_1_DECL:.*]]:2 = hlfir.declare %[[VAL_1]] {uniq_name = "_QFsingle_privatizationEy"} : (!fir.ref<f64>) -> (!fir.ref<f64>, !fir.ref<f64>)
+! CHECK:           omp.single   {
+! CHECK:             %[[X:.*]] = fir.alloca f32 {bindc_name = "x", pinned, uniq_name = "_QFsingle_privatizationEx"}
+! CHECK:             %[[X_DECL:.*]]:2 = hlfir.declare %[[X]] {uniq_name = "_QFsingle_privatizationEx"} : (!fir.ref<f32>) -> (!fir.ref<f32>, !fir.ref<f32>)
+! CHECK:             %[[Y:.*]] = fir.alloca f64 {bindc_name = "y", pinned, uniq_name = "_QFsingle_privatizationEy"}
+! CHECK:             %[[Y_DECL:.*]]:2 = hlfir.declare %[[Y]] {uniq_name = "_QFsingle_privatizationEy"} : (!fir.ref<f64>) -> (!fir.ref<f64>, !fir.ref<f64>)
+! CHECK:           %[[VAL_6:.*]] = fir.load %[[VAL_1_DECL]]#1 : !fir.ref<f64>
+! CHECK:           fir.store %[[VAL_6]] to %[[Y_DECL]]#1 : !fir.ref<f64>
+! CHECK:           fir.call @_QPbar(%[[X_DECL]]#1, %[[Y_DECL]]#1) fastmath<contract> : (!fir.ref<f32>, !fir.ref<f64>) -> ()
```
Comment (kiranchandramohan) [inline flang/test/Lower/OpenMP/single.f90:86]: Nit: Maintain the alignment.

File: flang/test/Lower/OpenMP/single.f90 [hunk 3]
```diff
@@ -0,0 +1,131 @@
+!RUN: %flang_fc1 -emit-hlfir -fopenmp %s -o - | FileCheck %s
+!RUN: bbc -emit-hlfir -fopenmp %s -o - | FileCheck %s
+
+!===============================================================================
+! Single construct
+!===============================================================================
+
+!CHECK-LABEL: func @_QPomp_single
+!CHECK-SAME: (%[[x:.*]]: !fir.ref<i32> {fir.bindc_name = "x"})
```
Comment (kiranchandramohan) [inline flang/test/Lower/OpenMP/single.f90:9]: Nit: Change `x` to a capital `X`, here and elsewhere.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

File: flang/test/Lower/OpenMP/single.f90 (hunk 1 above)
Comment (harishch4) [reply flang/test/Lower/OpenMP/single.f90]: Done

File: flang/test/Lower/OpenMP/single.f90 (hunk 2 above)
Comment (harishch4) [reply flang/test/Lower/OpenMP/single.f90]: Done

File: flang/test/Lower/OpenMP/single.f90 (hunk 3 above)
Comment (harishch4) [reply flang/test/Lower/OpenMP/single.f90]: Done

File: flang/test/Lower/OpenMP/single.f90 [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+! CHECK-LABEL: func.func @_QPsingle_privatization2(
+! CHECK-SAME:                                      %[[VAL_0:.*]]: !fir.ref<f32> {fir.bindc_name = "x"},
+! CHECK-SAME:                                      %[[VAL_1:.*]]: !fir.ref<f64> {fir.bindc_name = "y"}) {
+! CHECK:         %[[VAL_0_DECL:.*]]:2 = hlfir.declare %[[VAL_0]] {uniq_name = "_QFsingle_privatization2Ex"} : (!fir.ref<f32>) -> (!fir.ref<f32>, !fir.ref<f32>)
+! CHECK:         %[[VAL_1_DECL:.*]]:2 = hlfir.declare %[[VAL_1]] {uniq_name = "_QFsingle_privatization2Ey"} : (!fir.ref<f64>) -> (!fir.ref<f64>, !fir.ref<f64>)
+! CHECK:         omp.parallel   {
+! CHECK:           omp.single   {
+! CHECK:             %[[X:.*]] = fir.alloca f32 {bindc_name = "x", pinned, uniq_name = "_QFsingle_privatization2Ex"}
+! CHECK:             %[[X_DECL:.*]]:2 = hlfir.declare %[[X]] {uniq_name = "_QFsingle_privatization2Ex"} : (!fir.ref<f32>) -> (!fir.ref<f32>, !fir.ref<f32>)
```
[the remark this answers is not in the document]
Comment (harishch4) [reply flang/test/Lower/OpenMP/single.f90]: Done

File: flang/test/Lower/OpenMP/single.f90 [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  !$omp single private(x) firstprivate(y)
+  call bar(x, y)
+  !$omp end single
+end subroutine
+
+! CHECK-LABEL: func.func @_QPsingle_privatization2(
+! CHECK-SAME:                                      %[[VAL_0:.*]]: !fir.ref<f32> {fir.bindc_name = "x"},
+! CHECK-SAME:                                      %[[VAL_1:.*]]: !fir.ref<f64> {fir.bindc_name = "y"}) {
+! CHECK:         %[[VAL_0_DECL:.*]]:2 = hlfir.declare %[[VAL_0]] {uniq_name = "_QFsingle_privatization2Ex"} : (!fir.ref<f32>) -> (!fir.ref<f32>, !fir.ref<f32>)
+! CHECK:         %[[VAL_1_DECL:.*]]:2 = hlfir.declare %[[VAL_1]] {uniq_name = "_QFsingle_privatization2Ey"} : (!fir.ref<f64>) -> (!fir.ref<f64>, !fir.ref<f64>)
```
[the remark this answers is not in the document]
Comment (harishch4) [reply flang/test/Lower/OpenMP/single.f90]: Done

**Review by kiranchandramohan — APPROVED** [review]

**Comment by harishch4** [comment]
This current merge created a check-flang failure. A Fix(https://github.com/llvm/llvm-project/pull/70164) will be merged shortly.

**Pull request merged**